plugins/nvme/fu-plugin-nvme.c

```diff
@@ -0,0 +1,289 @@
+ */
+
+#include "config.h"
+
+#include <gudev/gudev.h>
+#include <json-glib/json-glib.h>
+#include <string.h>
+
+#include "fu-plugin.h"
+#include "fu-plugin-vfuncs.h"
+
+struct FuPluginData {
+	gchar			*nvme_fn;
+	GUdevClient		*gudev_client;
+};
+
+static gboolean
+fu_plugin_nvme_add_devfile (FuPlugin *plugin, const gchar *devfile, GError **error)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	JsonNode *node_root;
+	JsonObject *obj;
+	const gchar *nvme_fr;
+	const gchar *nvme_mn;
+	gint64 nvme_frmw;
+	g_autofree gchar *nvme_mn_fixed = NULL;
+	g_autofree gchar *stdout = NULL;
+	g_auto(GStrv) nvme_mn_split = NULL;
+	g_autoptr(FuDevice) dev = fu_device_new ();
+	g_autoptr(JsonParser) parser = json_parser_new ();
+	const gchar *argv[] = { data->nvme_fn, "id-ctrl", devfile,
+				"--output-format=json", NULL };
+
+	/* get JSON data */
+	if (!g_spawn_sync (NULL, (gchar **) argv, NULL, G_SPAWN_DEFAULT,
+			   NULL, NULL, &stdout, NULL, NULL, error))
+		return FALSE;
+	if (!json_parser_load_from_data (parser, stdout, -1, error)) {
+		g_prefix_error (error, "failed to parse node: ");
+		return FALSE;
+	}
+	node_root = json_parser_get_root (parser);
+	if (node_root == NULL) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INVALID_FILE,
+				     "no root node");
+		return FALSE;
+	}
+	obj = json_node_get_object (node_root);
+	if (obj == NULL) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INVALID_FILE,
+				     "no object");
+		return FALSE;
+	}
+
+	/* create device */
+	fu_device_add_flag (dev, FWUPD_DEVICE_FLAG_INTERNAL);
+	fu_device_add_flag (dev, FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_device_add_flag (dev, FWUPD_DEVICE_FLAG_REQUIRE_AC);
+	fu_device_set_platform_id (dev, devfile);
+	fu_device_set_summary (dev, "NVM Express Solid State Drive");
+	fu_device_add_icon (dev, "drive-harddisk");
+
+	/* get flags */
+	if (!json_object_has_member (obj, "frmw")) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INVALID_FILE,
+				     "no frmw member");
+		return FALSE;
+	}
+	nvme_frmw = json_object_get_int_member (obj, "frmw");
+	if ((nvme_frmw & 0x08) == 0)
+		fu_device_add_flag (dev, FWUPD_DEVICE_FLAG_NEEDS_REBOOT);
+
+	/* get manufacturer */
+	if (!json_object_has_member (obj, "mn")) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INVALID_FILE,
+				     "no mn member");
+		return FALSE;
+	}
+	nvme_mn = json_object_get_string_member (obj, "mn");
+
+	/* remove trailing spaces then add as the GUID */
+	nvme_mn_fixed = g_strchomp (g_strdup (nvme_mn));
+	fu_device_add_guid (dev, nvme_mn_fixed);
+
+	/* try to parse the mn into something more understandable */
+	nvme_mn_split = g_strsplit (nvme_mn_fixed, " ", -1);
+	if (g_strv_length (nvme_mn_split) == 2) {
+		if (g_strcmp0 (nvme_mn_split[1], "TOSHIBA") == 0 ||
+			   g_strcmp0 (nvme_mn_split[1], "TOS") == 0 ||
+			   g_strcmp0 (nvme_mn_split[1], "TO") == 0) {
```

Comment:
This sorta feels like it's going to grow into a pretty big list, how about just storing this in quirk data so it's easy to match?

Reply:
It's really only Toshiba that does it the wrong way around, but I do agree this kind of thing needs to go to the quirk database. Will fix.